src/app/qgsmaptooladdregularpolygon.cpp

```diff
@@ -0,0 +1,121 @@
+/***************************************************************************
+    qgsmaptooladdregularpolygon.h  -  map tool for adding regular polygon
+    ---------------------
+    begin                : July 2017
+    copyright            : (C) 2017
+    email                : lbartoletti at tuxfamily dot org
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmaptooladdregularpolygon.h"
+#include "qgscompoundcurve.h"
+#include "qgscurvepolygon.h"
+#include "qgsgeometryrubberband.h"
+#include "qgsgeometryutils.h"
+#include "qgsmapcanvas.h"
+#include "qgspoint.h"
+#include "qgisapp.h"
+#include "qgsstatusbar.h"
+
+QgsMapToolAddRegularPolygon::QgsMapToolAddRegularPolygon( QgsMapToolCapture *parentTool, QgsMapCanvas *canvas, CaptureMode mode )
+  : QgsMapToolCapture( canvas, QgisApp::instance()->cadDockWidget(), mode )
+  , mParentTool( parentTool )
+  , mTempRubberBand( nullptr )
+  , mRegularPolygon( QgsRegularPolygon() )
+{
+  clean();
+}
+
+QgsMapToolAddRegularPolygon::~QgsMapToolAddRegularPolygon()
+{
+  clean();
+}
+
+void QgsMapToolAddRegularPolygon::createNumberSidesSpinBox()
+{
+  deleteNumberSidesSpinBox();
+  mNumberSidesSpinBox = new QSpinBox();
+  mNumberSidesSpinBox->setMaximum( 99999999 );
+  mNumberSidesSpinBox->setMinimum( 3 );
+  mNumberSidesSpinBox->setPrefix( tr( "Number of sides: " ) );
+  mNumberSidesSpinBox->setValue( mNumberSides );
+  QgisApp::instance()->addUserInputWidget( mNumberSidesSpinBox );
+  mNumberSidesSpinBox->setFocus( Qt::TabFocusReason );
+}
+
+void QgsMapToolAddRegularPolygon::deleteNumberSidesSpinBox()
+{
+  if ( mNumberSidesSpinBox )
+  {
+    QgisApp::instance()->statusBarIface()->removeWidget( mNumberSidesSpinBox );
+    delete mNumberSidesSpinBox;
+    mNumberSidesSpinBox = nullptr;
```

Comment:
Maybe `std::unique_ptr` to avoid the manual deletion?

Reply:
Done.
Is std::make_unique (c++14) will be usable?
Wasn't for me.

Reply:
No, not available in master unfortunately